### ISSUE TITLE
some questions about do_range_projection function

### ISSUE THREAD

**Issue by mvish7** [DeyvidKochanov-TomTom/kprnet#8]
Hi,
Thanks for making this great work public. I'm having trouble understanding the do_range_projection() function. 

 I assume that the do_range_projection() function is based on the paper [scan-based semantic segmentation of lidar point clouds.](https://arxiv.org/pdf/2004.11803.pdf), and you have implemented a vectorized version of "Algorithm 1" from the paper. Please correct me if I'm wrong.

Below are my questions:

1. Algorithm 1 from the scan-based sem seg paper has Function GetRows(points). The GetRows() function has a line >>
` jump ← |φ[1 :]−φ[: −1]| > threshold` and as per my understanding you have implemented this specific line in [KPRNet code](https://github.com/DeyvidKochanov-TomTom/kprnet/blob/587a616b955b6be55b223e031e9ed39a8588d149/datasets/semantic_kitti.py#L154) as `new_raw = np.nonzero((proj_x[1:] < 0.2) * (proj_x[:-1] > 0.8))[0] + 1` .

   Could you please explain if these two lines are programmatically equivalent?? 

2. Could you please explain how the threshold values (i.e. 0.2 and 0.8) were chosen for the code line `new_raw = np.nonzero((proj_x[1:] < 0.2) * (proj_x[:-1] > 0.8))[0] + 1` ?? were they chosen as per the horizontal resolution of Velodyne LiDAR? 

3. If it is not too much then could you please explain the pseudo-code behind the following code lines from do_range_projection() function??
```
proj_x = 0.5 * (yaw / np.pi + 1.0)  # in [0.0, 1.0]
new_raw = np.nonzero((proj_x[1:] < 0.2) * (proj_x[:-1] > 0.8))[0] + 1
proj_y = np.zeros_like(proj_x)
proj_y[new_raw] = 1
proj_y = np.cumsum(proj_y)
proj_x = proj_x * W - 0.001
```

Thanks

**Comment by DeyvidKochanov-TomTom** [comment]
Hi,
its just a hack that exploits the fact that points in Kitti Odometry scans  still come in the order of their probe ID which we use to find where a new line starts.
This computes the azimuth angle for each point
```
yaw = -np.arctan2(scan_y, -scan_x) 
```
the next line normalizes the angle between 0,1 which used to compute the pixel coordinates of the point and find new lines
```
proj_x = 0.5 * (yaw / np.pi + 1.0)  
```
Here we find where a new line starts 

Firrst we find the points where the normalized azimuth angle is close to 1 (bigger than 0.8)  and the angle for the next point is close to 0 (smaller than 0.2).  The next point is then on a new line. 0.2 and 0.8 are not important just there to make sure it still works even if the lidar beam didn't return at the pixels close to the border
```
new_raw = np.nonzero((proj_x[1:] < 0.2) * (proj_x[:-1] > 0.8))[0] + 1
```
new_raw now contains indecies of the points which start a new line the next 3 lines just compute the y pixel coordinate base on that
```
proj_y = np.zeros_like(proj_x)
proj_y[new_raw] = 1
proj_y = np.cumsum(proj_y)
```


/  a small rant about why this is done starts here 
the projection method here is  just an approximation of the scan based projection from the paper. The paper uses Kitti Raw but the full mapping from Kitti Odometry (which is used by Semantic Kitti) to raw is not available. The mapping for the train and val is available and can be found in the README of the Kitti SDK but then some of the raw sequences are also not available. This is a problem because in the odometry dataset the scans are corrected for ego motion (probably using the data from the IMU?) and so the algorithm from the paper doesn't work. This was explained in the code of the scan base paper but I just checked and they have removed  it from the repo :thinking: . There are potentially other problems with the kitti lidar data. My cowerker has done some investigation you can find it in this repo https://github.com/olafbooij/ittik.

**Comment by mvish7** [comment]
Hi,
Thank you very much for this detailed explanation. I have LiDAR data from non-Velodyne made LiDAR and I could write create range images on the basis of your explanation.

**Comment by DeyvidKochanov-TomTom** [comment]
Hi,
cool! If it is a public dataset (ist it lyft?) you can make a PR for the dataloader!

**Comment by mvish7** [comment]
Hello,
The dataset is not public currently, we are still in the process of recording the dataset. Once the dataset is made public, I'll make sure to link the SDK here and make a PR.
